crates/rattler_networking/src/s3_middleware.rs

```diff
@@ -325,12 +340,78 @@ region = eu-central-1
                         .unwrap();
                     assert!(
                         presigned.to_string().contains("localhost:9000"),
-                        "Unexpected presigned URL: {:?}",
-                        presigned
+                        "Unexpected presigned URL: {presigned:?}"
                     );
                 })
             },
         )
         .await;
     }
+
+    #[rstest]
+    #[tokio::test]
+    #[serial]
+    async fn test_presigned_s3_request_custom_config() {
+        let temp_dir = tempdir().unwrap();
+        let aws_config = r#"
+        {
+            "s3://rattler-s3-testing/my-channel": {
+                "S3Credentials": {
+                    "access_key_id": "minioadmin",
+                    "secret_access_key": "minioadmin"
+                }
+            }
+        }
+        "#;
+        let credentials_path = temp_dir.path().join("credentials.json");
+        std::fs::write(&credentials_path, aws_config).unwrap();
+        let s3 = S3::new(
+            S3Config::Custom {
+                endpoint_url: Url::parse("http://localhost:9000").unwrap(),
+                region: "eu-central-1".into(),
+                force_path_style: true,
+            },
+            AuthenticationStorage::from_file(credentials_path.as_path()).unwrap(),
+        );
+
+        let presigned = s3
+            .generate_presigned_s3_url(
+                Url::parse("s3://rattler-s3-testing/my-channel/noarch/repodata.json").unwrap(),
+            )
+            .await
+            .unwrap();
+        assert_eq!(
+            presigned.path(),
+            "/rattler-s3-testing/my-channel/noarch/repodata.json"
+        );
+        assert_eq!(presigned.scheme(), "http");
+        assert_eq!(presigned.host_str().unwrap(), "localhost");
+        assert!(presigned.query().unwrap().contains("X-Amz-Credential"));
+    }
+
+    #[rstest]
+    #[tokio::test]
+    #[serial]
+    async fn test_presigned_s3_request_public_bucket() {
```

Comment:
we should also have a test that ensures public buckets work with `S3Config::FromAWS`

Reply:
Done